server/map_tree_updater.go

```diff
@@ -106,11 +110,20 @@ func (t *mapTreeUpdater) update(ctx context.Context, tx storage.MapTreeTX, upd [
 }
 
 type txAccessor struct {
+	// hashes is a cache of node hashes that Get returns directly instead of
+	// calling GetMerkleNodes. Used only if preload is true.
+	hashes  map[tree.NodeID2][]byte
+	preload bool
+
 	tx  storage.MapTreeTX
 	rev int64
 }
 
 func (t txAccessor) Get(ctx context.Context, ids []tree.NodeID2) (map[tree.NodeID2][]byte, error) {
```

Comment:
This feels like a completely different way of doing Gets.  
Would it make sense to have a *different* txAccessor for the cached reads?
The one off - ifs make me nervous that this will be harder to test.

Reply:
Don't worry about the tests, both cases are covered already (see codecov report).

Could do the splitting, but the `Set` method is shared, so I would need some extra machinery around it to have 2 half-identical implementations. Also, there is a high chance that the preload mode will be removed.

Plus, this accessor is not in the final shape, I'm changing it in #1932 to use tileset. How about leaving a TODO to potentially break this type down when some unclear bits are settled?

Reply:
Added TODO.